File: src/main/java/dev/sky_lock/pocketlifevehicle/car/Speed.java
package dev.sky_lock.pocketlifevehicle.car;

import java.math.BigDecimal;

/**
 * @author sky_lock
 */

class Speed {
    private final static BigDecimal acceleration = new BigDecimal("0.0085");
    private final static BigDecimal frictionalDeceleration = new BigDecimal("0.010");
    private final static BigDecimal magnification = new BigDecimal("0.85");
    private BigDecimal exactSpeed = BigDecimal.ZERO;

    void accelerate() {
        exactSpeed = exactSpeed.add(acceleration);
    }

    void frictionalDecelerate() {
        exactSpeed = exactSpeed.subtract(acceleration);
    }

    void decelerate() {
        exactSpeed = exactSpeed.subtract(acceleration.add(frictionalDeceleration));
    }

    void decrease() {
        exactSpeed = exactSpeed.multiply(magnification);
    }

    boolean isPositive() {
        return exactSpeed.compareTo(BigDecimal.ZERO) > 0;
    }

    boolean isNegative() {
        return exactSpeed.compareTo(BigDecimal.ZERO) < 0;
    }

    float approximate() {
        return exactSpeed.setScale(3, BigDecimal.ROUND_HALF_UP).floatValue();
    }

    boolean isApproximateZero() {
        return approximate() == 0.0F;
    }

    float exact() {
        return exactSpeed.floatValue();
    }

    void zeroize() {
        exactSpeed = BigDecimal.ZERO;
    }

}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/vehicle/Speed.kt
package dev.sky_lock.pocketlifevehicle.vehicle

import java.math.BigDecimal

/**
 * @author sky_lock
 */
class Speed {
    private var exactSpeed = BigDecimal.ZERO

    val isPositive: Boolean
        get() = exactSpeed > BigDecimal.ZERO

    val isNegative: Boolean
        get() = exactSpeed < BigDecimal.ZERO

    fun accelerate() {
        exactSpeed += acceleration
    }

    fun frictionalDecelerate() {
        exactSpeed -= acceleration
    }

    fun decelerate() {
        exactSpeed -= acceleration + frictionalDeceleration
    }

    fun decrease() {
        exactSpeed *= magnification
    }

    fun approximate(): Float {
        return exactSpeed.setScale(3, BigDecimal.ROUND_HALF_UP).toFloat()
    }

    val isApproximateZero: Boolean
        get() = approximate() == 0.0f

    fun exact(): Float {
        return exactSpeed.toFloat()
    }

    fun zeroize() {
        exactSpeed = BigDecimal.ZERO
    }

    companion object {
        private val acceleration = BigDecimal("0.0085")
        private val frictionalDeceleration = BigDecimal("0.010")
        private val magnification = BigDecimal("0.85")
    }
}